### ISSUE TITLE
Push official images to quay.io, in addition to Docker Hub

### ISSUE THREAD

**Issue by aditya-konarde** [jaegertracing/jaeger#2498]
# Requirement - what kind of business use case are you trying to solve?

High availability and resiliency of Jaeger in production environments

## Problem - what in Jaeger blocks you from solving the requirement?

Currently, the official images for jaeger are on docker hub. Unless one has a docker paid account, image pulls will be rate-limited and pulls may fail at run-time, causing infrastructure issues.

## Proposal - what do you suggest to solve the problem or improve the existing situation?

$title: Push official images to quay.io, in addition to Docker Hub

We already have an organization where we push the jaeger operator https://quay.io/repository/jaegertracing/jaeger-operator

I propose that we also push all the other images to quay.io, in addition to the docker hub repository here: https://hub.docker.com/u/jaegertracing

## Any open questions to address

None

**Comment by jpkrohling** [comment]
I'm OK with this idea. @yurishkuro, what do you think?

**Comment by yurishkuro** [comment]
_in addition_ sounds fine.

**Comment by aditya-konarde** [comment]
Thanks, I'll start to work on it this week :)

**Comment by aebirim** [comment]
Hi @aditya-konarde, @jpkrohling 

I'm an Outreachy applicant and would like to tackle this issue.

**Comment by aditya-konarde** [comment]
Sounds good. I haven't gotten the time to work on this yet, so please go ahead @aebirim :)

**Comment by aebirim** [comment]
Hi @aditya-konarde will do!

**Comment by aebirim** [comment]
Hi @jpkrohling I've uploaded the 41 images to the quay.io platform. Some of the latest tags didn't work so used the next most recent one. 

I've sent you a message on gitter.

**Comment by jpkrohling** [comment]
@aebirim, ideally, we would push the images to both places during the release process. It's currently called during a Travis build here:

https://github.com/jaegertracing/jaeger/blob/0f88343172c0bb16ad5c1805d18e20d59176ea7e/.travis.yml#L73-L74

Take a look and see if you can follow what it's doing end-to-end, but I think you might need to change only two places:

1. `scripts/travis/upload-all-docker-images.sh`, to call the `upload-to-docker.sh` for both `docker.io` and for `quay.io`
1. `scripts/travis/upload-to-docker.sh`, to accommodate the case where images are pushed to `quay.io`

**Comment by morlay** [comment]
could we push github contrainer registry `ghcr.io/jaegertracing/*` too?

**Comment by yurishkuro** [comment]
I don't mind having images pushed to more than one places, but it would be better to do it asynchronously, i.e. our build does a push to a single place (Docker Hub) other repositories sync from there. Adding more pushes to the main build will only make it less reliable.

**Comment by jpkrohling** [comment]
> and other repositories sync from there

Do you know what can be used to do this? I don't remember seeing sync to/from Quay/GitHub/Docker.

**Comment by rsvoboda** [comment]
@jpkrohling / @aebirim could be this effort refreshed please, I see PR is staled for ~1 month.

Asking because we are suffering by dockerhub rate limit apocalypse when running MicroProfile OpenTracing related tests.
We are able to propagate pull secret on OpenShift 4 to overcome this trouble, but on OCP 3.11 the same solution can't be applied.

We are looking at image alternatives available on Quay.io and jaeger image is one of the images pulled from docker hub.

**Comment by jpkrohling** [comment]
@rsvoboda is there a specific version that you are using? I could push them to quay manually for the time being.

**Comment by rsvoboda** [comment]
Hi Juca. No rush atm, we had applied workaround for current release, we will have new testing during January.
It would be ideal to have quay mirror by that time. If not, I would ask you for adhoc push ;)

**Comment by rsvoboda** [comment]
Hi @jpkrohling, any update on quay.io effort? If no, I will look into the image versions I would like to pull from quay.io. Thanks.

**Comment by jpkrohling** [comment]
I don't have news at the moment. There is a PR that would make it possible to send the same image to other repositories, but it hasn't been merged either. For now, do let me know which images you'd need there and I'll push them manually.

**Comment by rsvoboda** [comment]
We use `jaegertracing/all-in-one`, we do not specify concrete version though, just pulling the latest.
We can depend on concrete version for now, for example `1.21.0`. Hope this issue gets resolved soon.

Did some searching on quay.io (https://quay.io/search?q=jaegertracing) and some people do their own "mirrors".

**Comment by jpkrohling** [comment]
`quay.io/jaegertracing/all-in-one:1.21.0` is now available now

**Comment by rsvoboda** [comment]
https://quay.io/jaegertracing/all-in-one gives 403 Unauthorized, isn't the repository set to private?

**Comment by rsvoboda** [comment]
I can access ^^^ now, thanks

**Comment by ehsavoie** [comment]
could a latest be defined so we don't have to specify the version ?

**Comment by jpkrohling** [comment]
Done. Note that until we have a proper publishing mechanism for quay, there's a risk that `latest` won't necessarily match the latest we have in Docker Hub

**Comment by achandras** [comment]
Hi, we're also getting 403 on https://quay.io/jaegertracing/jaeger-agent, could that be opened up too? Thanks

**Comment by jpkrohling** [comment]
> Hi, we're also getting 403 on https://quay.io/jaegertracing/jaeger-agent, could that be opened up too? Thanks

That one doesn't actually exist. Which version do you need there? I can try to do a manual push.

@Ashmita152, do you think it's easy now to add this to the CI? I know you've done some work around this lately.

**Comment by achandras** [comment]
Thanks for the quick response @jpkrohling, we would like to use 1.21.0 from the quay repo if possible

**Comment by Ashmita152** [comment]
> @Ashmita152, do you think it's easy now to add this to the CI? I know you've done some work around this lately.

Sure, I will work on it. Yes, it should be easy.

**Comment by jpkrohling** [comment]
> Thanks for the quick response @jpkrohling, we would like to use 1.21.0 from the quay repo if possible

Sorry for the delay, I missed the notification. The linked PR should make publishing happen automatically to both docker.io and quay.io. For 1.21.0, I just pushed the following images there: jaeger-agent, jaeger-collector, jaeger-ingester, jaeger-query, all-in-one

**Comment by yurishkuro** [comment]
We probably need to pre-create all the repositories on quay:
```
+ docker push quay.io/jaegertracing/example-hotrod
Using default tag: latest
The push refers to repository [quay.io/jaegertracing/example-hotrod]
a93b60b49a79: Preparing
unauthorized: access to the requested resource is not authorized
Error: Process completed with exit code 1.
```

**Comment by jpkrohling** [comment]
The account had permissions to push only to the jaeger-operator repository. I just changed it to be able to push to all repos.

**Comment by yurishkuro** [comment]
Doesn't help, I think repos must be manually created on quay first before the bot can push to them.

**Comment by jpkrohling** [comment]
Typically not, but for the robot account, it does seem to be the case.

**Comment by jpkrohling** [comment]
I'm creating the missing repos, but won't be done until probably tomorrow.

**Comment by jpkrohling** [comment]
Done, looks like the build now works.

**Comment by albertteoh** [comment]
I believe this issue is addressed in #2783; closing for now.
